### ISSUE TITLE
Keras Model produces nan on fit

### ISSUE THREAD

**Issue by arose13** [tensorflow/tensorflow#40651]
**System information**
- Google Colab Python 3

**Bug Description**
I have a confusing problem. Before I call `.fit()` my model can produce output from inputs (see below).

```python

>>> print(f0.predict(x))
>>> print(f1.predict([x, x]))

[[16.913166 ]
 [ 6.2363415]
 [10.625664 ]
 ...
 [13.86934  ]
 [15.4675865]
 [16.353233 ]]
[[-0.94484174]
 [-1.0965291 ]
 [-0.829823  ]
 ...
 [-1.0362396 ]
 [-0.18886864]
 [-0.3576672 ]]
```

But after calling fit all I get are nans
```python
>>> print(f0.predict(x))
>>> print(f1.predict([x, x]))

[[16.913166 ]
 [ 6.2363415]
 [10.625664 ]
 ...
 [13.86934  ]
 [15.4675865]
 [16.353233 ]]
[[nan]
 [nan]
 [nan]
 ...
 [nan]
 [nan]
 [nan]]
```

Here is the code for generating the model. Error is _not_ dependent on the data as I can reproduce this with any sklearn dataset as an input.

```python
def create_base_model(p, units=20):
    # Specs
    input_layer = k.layers.Input((p,))

    penultimate_layer = k.layers.Dense(units, activation='elu')(input_layer)
    penultimate_layer = k.layers.Dense(p)(penultimate_layer)

    training_output = k.layers.Dense(1)(penultimate_layer)

    # Models
    boost_model = k.Model(inputs=input_layer, outputs=penultimate_layer)
    training_model = k.Model(inputs=input_layer, outputs=training_output)

    # Compile and export
    training_model.compile(optimizer='sgd', loss='mse')
    return training_model, boost_model


def create_staged_model(p: int, model: k.Model, units=20):
    # Freeze prior model
    for layer_i in model.layers:
        layer_i.trainable = False

    # Specs
    input_layer = k.layers.Input((p,))

    penultimate_layer = k.layers.concatenate([model.output, input_layer], axis=-1)
    penultimate_layer = k.layers.Dense(units, activation='elu')(penultimate_layer)
    penultimate_layer = k.layers.Dense(p)(penultimate_layer)

    training_output = k.layers.Dense(1)(penultimate_layer)

    # Models
    boost_model = k.Model(inputs=[model.input, input_layer], outputs=penultimate_layer)
    training_model = k.Model(inputs=[model.input, input_layer], outputs=training_output)

    # Compile and export
    training_model.compile(optimizer='sgd', loss='mse')
    return training_model, boost_model


fit_kwargs = dict(
    epochs=50, 
    validation_split=0.1, 
    callbacks=[
        k.callbacks.ReduceLROnPlateau(factor=.5, patience=5, min_lr=1e-6),
        k.callbacks.EarlyStopping(patience=10, restore_best_weights=True)
    ]
)


f0, _ = create_base_model(p)
f0.fit(x, y, **fit_kwargs)  # Works fine

f1, _ = create_staged_model(p, model0)
f1.fit([x, x], y, **fit_kwargs)  # Breaks on fit
```

Note that this nan issue occurs immediately
```
Epoch 1/50
282/282 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan - lr: 0.0100
Epoch 2/50
282/282 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan - lr: 0.0100
Epoch 3/50
282/282 [==============================] - 0s 1ms/step - loss: nan - val_loss: nan - lr: 0.0100
Epoch 4/50
282/282 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan - lr: 0.0100
Epoch 5/50
282/282 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan - lr: 0.0100
Epoch 6/50
282/282 [==============================] - 0s 1ms/step - loss: nan - val_loss: nan - lr: 0.0050
Epoch 7/50
282/282 [==============================] - 0s 1ms/step - loss: nan - val_loss: nan - lr: 0.0050
Epoch 8/50
282/282 [==============================] - 0s 2ms/step - loss: nan - val_loss: nan - lr: 0.0050
Epoch 9/50
282/282 [==============================] - 0s 1ms/step - loss: nan - val_loss: nan - lr: 0.0050
```

**Comment by ravikyram** [comment]
@arose13 

I have tried in colab with TF version 2.2 and i am seeing the below error messaage.(N`ameError: name 'p' is not defined`).Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/c5b3678e7d8743f2301dca44e74d283b/untitled56.ipynb).Please, help me with reproducible code, so it helps us in localzing the issue faster.Thanks!

**Comment by arose13** [comment]
I have no idea if my changed affected your gist so I'm sharing my URL here of your colab. [https://colab.research.google.com/gist/arose13/a580640c042278b82d6d7dc424070592/untitled56.ipynb](https://colab.research.google.com/gist/arose13/a580640c042278b82d6d7dc424070592/untitled56.ipynb )

I added summary prints so you can a little more about what is going on. I'm also having a different error than in your Colab but the observed behaviour is still the same (and strange).

**Comment by yixingfu** [comment]
@arose13 

Use tensorflow.keras rather than keras will fix the problem.

**Comment by yixingfu** [comment]
@arose13  You are right (but I think I did see that worked before I replied above. Not sure if I modified anything else at that time.)

If f0 is not trained, training f1 does not have problem.  Also if I change optimizer from 'sgd' to 'adam' it does not have nan.  Maybe it is a sgd problem caused by the training result of f0?

**Comment by arose13** [comment]
Incredibly clever fix @yixingfu !

I am so incredibly confused, do you have any intuition why this works?

f0 appears normal and I can `.predict()` etc, so why would something as arbitrary as the 'sgd' -> 'adam' work (it has worked). 

**Note 1**: Only changed the second model `f1` to 'adam' fixes it. Changing only `f0` does not. This continues to make me believe that somehow the problem is with how `f1` is created (created by `create_staged_model()`).

**Note 2**: The reason why it is important is that I must train the staged models (eg `f1` ) with stochastic gradient descent.

**Comment by sachinprasadhs** [comment]
I did not face any issues while running your code in Tf Nightly 2.6, please find the gist [here](https://colab.research.google.com/gist/sachinprasadhs/b412e966a3fb4b6c2dbd202c36563713/40651.ipynb) and confirm us if the issue is fixed. Thanks!
